tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15138,6 +15138,63 @@ static bool nodeInputIsCompatible(DXIL::NodeIOKind IOType,
   }
 }
 
+// Diagnose input node record to make sure it has exactly one SV_DispatchGrid
+// semantics. Recursivelly walk all fields on the record and all of its base
+// classes/structs
+void DiagnoseMustHaveOneDispatchGridSemantics(Sema &S,
+                                              CXXRecordDecl *InputRecordDecl,
+                                              SourceLocation &DispatchGridLoc,
+                                              bool &Found) {
+  // Iterate over fields of the input record struct
+  for (auto FieldDecl : InputRecordDecl->fields()) {
+    // Check if any of the fields have SV_DispatchGrid annotation
+    for (const hlsl::UnusualAnnotation *it :
+         FieldDecl->getUnusualAnnotations()) {
+      if (it->getKind() == hlsl::UnusualAnnotation::UA_SemanticDecl) {
+        const hlsl::SemanticDecl *sd = cast<hlsl::SemanticDecl>(it);
+        if (sd->SemanticName.equals("SV_DispatchGrid")) {
+          if (!Found) {
+            Found = true;
+            DispatchGridLoc = it->Loc;
+          } else {
+            // There should be just one SV_DispatchGrid in per record struct
+            S.Diags.Report(
+                it->Loc,
+                diag::err_hlsl_dispatchgrid_semantic_already_specified);
+            S.Diags.Report(DispatchGridLoc, diag::note_defined_here)
```

Comment:
Strangely, `fields()` must iterate over fields in base classes in the reverse order than what I would expect, leading to these diagnostics being reversed from expectation.  It places the duplicate diag on the base field and the note on the derived field.  That's weird, because you would expect it to be the other way around, which would be more helpful.  There's probably another way to iterate bases and fields such that they go from bases first, but it would require more complicated iteration code.

Reply:
Oh, nevermind, I hadn't read the rest of the code that handles bases after the fields here.  All you have to do is swap the ordering to iterate bases first.

Reply:
Addressed in PR #5844